### PR TITLE
MM-52805 - Calls: Add calls ringing & settings panel option

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- For the new DM/GM calls ringing feature, we need a way to turn the calls ringing on/off, and a way to choose the ring tone.
- An option would be to add a pluggable to the desktop notifications panel, but we would need to create a lot of plumbing to get the proper save/cancel flow set up. This was easier. It is only shown when the calls plugin (>= `v0.17.0`) is installed. That is the version we're shooting for to implement ringing. If we miss it, I'll submit a revised version string before. (aside: open to putting the version string as a constant somewhere if you'd prefer, I just saw that the desktop version string is inline, so following the style).
- Note that the screenshot of the `Never` condition looks the same as main -- the extra `hr`s are there from before. Also, the `Notification sound` dropdown saves on every change, regardless of selecting `Save` or `Cancel`. I gather this isn't a big UX issue if it hasn't been fixed by now?

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-52805
- Designs: https://www.figma.com/file/GjmY3V0G6K88FruOcuMP7G/MM-51513-Ringing-in-DMs%2C-GMs?type=design&node-id=918-212836&t=mkwlPlViilS6mJfd-0

#### Screenshots
image
image

### With the `Never` notifications 
image

#### Release Note

```release-note
Calls: Ringing sounds can be enabled/disabled and selected in the Desktop Notifications preferences panel.
```

### CODE REVIEW

File: webapp/channels/src/components/user_settings/notifications/index.ts [hunk 1]
```diff
@@ -12,17 +12,21 @@ import {ActionFunc} from 'mattermost-redux/types/actions';
...
 import UserSettingsNotifications, {Props} from './user_settings_notifications';
+import {suitePluginIds} from 'utils/constants';
+import semver from 'semver';
 
 function mapStateToProps(state: GlobalState) {
     const config = getConfig(state);
 
     const sendPushNotifications = config.SendPushNotifications === 'true';
     const enableAutoResponder = config.ExperimentalEnableAutomaticReplies === 'true';
+    const isCallsEnabled = state.plugins.plugins[suitePluginIds.calls] && semver.gte(state.plugins.plugins[suitePluginIds.calls].version || '', '0.17.0');
```
[the remark this answers is not in the document]
Comment (cpoile) [reply webapp/channels/src/components/user_settings/notifications/index.ts]: I put the version as an optional parameter in the selector, that way we get both without needing a util function. Sound good?

**Review by hmhealey — APPROVED** [review]
This looks good to me. I do wonder if more of it should be done via the plugin API (like exposing a way for plugins to add user settings), but I'm still unsure if Calls is going to always remain a plugin or if we're planning on integrating it into server/apps directly

**Comment by cpoile** [comment]
@hmhealey Adding a Pluggable was my first (and preferred) option, but I realized it would require recreating the complex architecture that we use with the System Console to allow plugins to review changes before allowing a save, canceling the save, etc. We could do that, if we wanted, but it would be a much larger effort than just adding a setting like we have here.

File: webapp/channels/src/components/user_settings/notifications/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (streamer45) [reply webapp/channels/src/components/user_settings/notifications/index.ts]: Yep, perfect :100:

**Comment by hmhealey** [comment]
Yeah, and the settings modal certainly isn't written in a way that makes it easy to expose or add parts to it. There's so much raw HTML and Bootstrap in there that it's not easy to extend. Perhaps whenever we finally get around to redoing those modals, we can do them in a way that's more easily extensible and data-driven

File: webapp/platform/types/src/users.ts [hunk 2]
```diff
@@ -20,6 +21,7 @@ export type UserNotifyProps = {
     channel: 'true' | 'false';
     mention_keys: string;
     desktop_notification_sound?: 'Bing' | 'Crackle' | 'Down' | 'Hello' | 'Ripple' | 'Upstairs';
+    calls_notification_sound?: 'Dynamic' | 'Calm' | 'Urgent' | 'Cheerful';
```
Comment (streamer45) [inline webapp/platform/types/src/users.ts:24]: If we follow the existing convention should this be `calls_desktop_notification_sound` or `desktop_calls_notification_sound`? Otherwise we may as well drop `desktop` altogether but it depends on the answer to my previous question I guess.

File: webapp/platform/types/src/users.ts [hunk 3]
```diff
@@ -11,6 +11,7 @@ import {IDMappedObjects, RelationOneToMany, RelationOneToManyUnique, RelationOne
 export type UserNotifyProps = {
     desktop: 'default' | 'all' | 'mention' | 'none';
     desktop_sound: 'true' | 'false';
+    calls_desktop_sound: 'true' | 'false';
```
[the remark this answers is not in the document]
Comment (cpoile) [reply webapp/platform/types/src/users.ts:14]: Just confirmed with Matt that yes, mobile should have separate on/off preferences. As for sound, see below.

File: webapp/platform/types/src/users.ts (hunk 2 above)
Comment (cpoile) [reply webapp/platform/types/src/users.ts:24]: At first I chose this because `calls_desktop_notification_sound` was just too long. But now (see: https://community.mattermost.com/core/pl/uqz5m6czqtbquy384goyq6ifie) I think the consensus is that the sound will be shared across desktop/web and mobile, so the name reflects that it's for both. If we think we might have a separate one for mobile in the future, we could name the mobile one `calls_notification_sound_mobile` to differentiate (if it's ever needed). 
tldr: imo, keep this and if we do need a separate mobile, make that one a more specific name.

**Review by streamer45 — APPROVED** [review]

File: webapp/platform/types/src/users.ts (hunk 2 above)
Comment (streamer45) [reply webapp/platform/types/src/users.ts:24]: Sounds good, thanks for confirming 👍

**Comment by cpoile** [comment]
Hey @DHaussermann , ping me when you're ready to take a look at this -- we'll have to spend some time getting you set up to test calls before this PR will work correctly.

**Comment by cpoile** [comment]
@DHaussermann should be good to go, thanks!

**Comment by DHaussermann** [comment]
Hi @cpoile I did do a round of testing on this before you pulled in master. 

One minor issue is with the tone continuing to play after the modal closes. If I select a tone and quickly close the modal with the save button the sound of the selected tone stops immediately as expected.

However if I close the modal with the `X` button instead the tone will continue to play for the full cycle. 

I tried to test again with the latest commit today but I'm having deployment issues. I'll follow-up if I can't get them sorted.

**Comment by cpoile** [comment]
> However if I close the modal with the `X` button instead the tone will continue to play for the full cycle.

Thanks @DHaussermann for this catch. I fixed that, plus another one (turning calls notifications "off" while a ring was playing). Please take another look!

**Comment by DHaussermann** [comment]
I did another round of testing on this. 
- The sound will stop playing when: 
  - The drop down is switched to a new sound
  - Save is clicked
  - Cancel is clicked
  - `X` is clicked to close the mdal
  - `ESC` keystroke closes the modal
- Tested other edge cases such as mobile web view
- Saving works mostly as expected.    

@cpoile the only other small quirk I see that is a bit counter intuitive is when the selected sound saves. If you make changes then click cancel, changed to the the state of the On/Off radio are discarded as expected but, changes to the sound drop down are preserved. Possibly this gets saved as soon as a selection is made.
 
This is not just for the new Notification sound for incoming calls but also for the existing Notification sound section. 
image
You can modify both fields and then click cancel. You'll see when you return that the changes to the sound selection are preserved.

Please advise if I should create a new issue to prioritize this separately as it may be lower priority than other work.

**Comment by cpoile** [comment]
@DHaussermann Yep, I noted that in the third bullet point in the summary. Sure, submit a ticket and we'll see if it's important (doubt it :) ). Thanks for the thorough review!

**Review by DHaussermann — APPROVED** [review]
Tested and passed as per test summary above. 

There is an existing piece that is not UX bug per-say but a bit unintuitive. It's very minor but I created https://mattermost.atlassian.net/browse/MM-53166 to triage it.

LGTM!

**Comment by amyblais** [comment]
@cpoile Please help merge this PR for today's v8.0 feature complete deadline.

**Comment by RickyTR** [comment]
I use server ver. 7.10.2 and just updated to calls plugin 0.17 but calls notification options are not visible in desktop app (5.4.0). Is there something that I nedd to do other than switch on 'enable call ringing' in plugin stetings? 
Thanks a lot.

**Comment by cpoile** [comment]
@RickyTR It may be that you need to have v8.0.0 of Mattermost server. Hope that helps?

**Comment by RickyTR** [comment]
> @RickyTR It may be that you need to have v8.0.0 of Mattermost server. Hope that helps?

Oh sorry. I thought I need only to update the plugin. Thanks a lot.
